### ISSUE TITLE
Both junit-jupiter-engine and a custom registered engine are run in IDEA

### ISSUE THREAD

**Issue by igorstojanovski** [junit-team/junit5#1372]
## Overview

I am trying to create my own implementation of TestEngine, MyCustomExecutionEngine. I want to re-use the discovery mechanism from JupiterTestEngine and just do a custom execute() implementation. Hence, I want the test junit-jupiter-engine present as a dependency. But, I don't want the junit-jupiter-engine to run, I want only my custom engine to run.

## Problem

- I register MyCustomExecutionEngine in META-INF/services/org.junit.platform.engine.TestEngine 
- I run a test class from IDEA. The test class and my custom engine implementation are in the same project. 
- Because the implemented methods in MyCustomExecutionEngine are empty I get this warning (twice):
> WARNING: TestEngine with ID 'my-custom-engine' failed to discover tests
- But then the tests are run after all.
- When debugging, in DefaultLauncher#discoverRoot (called twice) I can see that testEngines iterable contains three engines: MyCustomExecutionEngine, JupiterTestEngine, and VintageTestEngine. 
- If I completely remove junit-jupiter-engine as a dependency the problem is solved. 

 ## Versions

- Junit 5.1.0
- IDEA 2017.3.4

**Comment by sbrannen** [comment]
The JUnit Platform provides first-class support for excluding test engines via `org.junit.platform.launcher.EngineFilter.excludeEngines(String...)`.

That filter can be supplied to the `LauncherDiscoveryRequestBuilder` that is used to build the `LauncherDiscoveryRequest` which is turn handed off to the `Launcher`.

The support is there, but as for whether or not you can achieve this via the UI in IntelliJ IDEA depends on whether IDEA supports including/excluding engines.

In summary, I am closing this issue as _works as designed_ since it really depends on what is implemented in IDEA.

p.s. if you're using the JUnit Platform Gradle Plugin or the `ConsoleLauncher`, you have support for including/excluding engines there.

**Comment by marcphilipp** [comment]
Sam is of course right: using an `EngineFilter` is a possible solution. However, Sam didn't know that I asked you to create an issue [on Twitter](https://twitter.com/marcphilipp/status/985083691205619712) because I was curious why you wanted to create your own engine that reuses the Jupiter engine.

@igorstojanovski Can you please provide a little more background on that?

**Comment by igorstojanovski** [comment]
@sbrannen  that, of course, worked perfectly fine. I knew I was missing something. However, that does not solve my IDEA problem. I couldn't find any way to specify which test engine(s) to be used to run the tests. And I still need to try this with Maven as well.

@marcphilipp  I just want to do change the way tests are executed, not the way they are discovered. And because I have no specific requirements for the discovery phase, I don't want to implement the TestEngine#discover method from scratch. Basically, I just want to copy-paste the discover method from JupiterTestEngine.

**Comment by smoyer64** [comment]
@marcphilipp - My original JupiterSuiteEngine was a wrapper around the regular Jupiter engine.  I excluded the Jupiter engine and called into the Jupiter engine to discover and execute the tests.

**Comment by sbrannen** [comment]
> However, that does not solve my IDEA problem. I couldn't find any way to specify which test engine(s) to be used to run the tests. And I still need to try this with Maven as well.

If you want IDEA to support filters for test engines, you would need to open a ticket in their issue tracker requesting that support.'

AFAIK, Eclipse does not support filtering of engines either.

**Comment by sbrannen** [comment]
> And I still need to try this with Maven as well.

The JUnit Platform Maven Surefire provider currently does not support filtering by test engine.

If you would like it to, feel free to open a new issue in this GitHub repository requesting that.

**Comment by sbrannen** [comment]
>  I just want to do change the way tests are executed

Can you please expound on exactly what it is you want to change?

**Comment by igorstojanovski** [comment]
@sbrannen I want to run the tests in parallel. The run is made based on user configuration who enters number of threads and whether to run classes, methods in classes or all methods in parallel.

**Comment by akozlova** [comment]
And similar issue bites me if I want my tests to run on EDT: I want to reuse the discovery but don't need to run with the engine. I still hope that there would be a solution for that

Apart from that. From IDEA point of view:
It's impossible to detect what filter would user want based on the opened class in the editor, thus configuring such things in the IDE would be complicated and I doubt that it's the correct way to deal with the problem. I'd rather add some override marker to the engine so once in the classpath, it will totally replace the standard engine. This way all the work would be done by engine's author and all the clients won't need to configure the same filter over and over again. Hope this makes sense. Thanks

**Comment by sbrannen** [comment]
> I want to run the tests in parallel. The run is made based on user configuration who enters number of threads and whether to run classes, methods in classes or all methods in parallel.

@igorstojanovski, wouldn't you rather wait on the official support for parallel execution coming in #60?

**Comment by sbrannen** [comment]
@akozlova, what is _EDT_?

**Comment by akozlova** [comment]
Event Dispatch Thread?

**Comment by sbrannen** [comment]
> From IDEA point of view: It's impossible to detect what filter would user want based on the opened class in the editor, thus configuring such things in the IDE would be complicated and I doubt that it's the correct way to deal with the problem. 

Using an `EngineFilter` is not the correct solution for an engine that wants to _replace_ an existing engine that happens to be on the classpath. That is of course another story. 😉 

However, build tools and IDEs _should_ actually support filtering by engine IDs just like they support filtering by tags.

**Comment by sbrannen** [comment]
> I'd rather add some override marker to the engine so once in the classpath, it will totally replace the standard engine. This way all the work would be done by engine's author and all the clients won't need to configure the same filter over and over again. Hope this makes sense.

Yes, that of course makes sense, and I think it would actually be a robust mechanism for achieving that... in case the JUnit Team decides to support that.

**Comment by sbrannen** [comment]
Replying to myself...

> However, build tools and IDEs _should_ actually support filtering by engine IDs just like they support filtering by tags.

Rationale: developers who are running tests within a given project may well want to exclude certain engines when executing _test suites_.

This can already be achieved via the `@IncludeEngines` and `@ExcludeEngines` annotations in the JUnit Platform Suite API. 

Unfortunately, those annotations are currently only _supported_ via the `JUnitPlatform` runner for JUnit 4 infrastructure. A better solution will be available once #744 is resolved.

See also: [suite tag](https://github.com/junit-team/junit5/labels/theme%3A%20suites)

**Comment by igorstojanovski** [comment]
> wouldn't you rather wait on the official support for parallel execution coming in #60?

@sbrannen It might be of help, but it probably won't do the whole job for me? "Run tests in parallel" is the short version. The long is that I need to group the tests first according to the setup and resources they use. Then I run in parallel only the tests that belong to the same group. But, I run the groups in sequence, one after the other. 

Hope that made sense.

**Comment by marcphilipp** [comment]
One idea I had was to define a default filter configuration in `junit-platform.properties` which is used in case no filter is part of a certain type, e.g. `EngineFilter`, is present in the `DiscoveryRequest`.

@junit-team/junit-lambda Thoughts?

**Comment by marcphilipp** [comment]
@igorstojanovski Is this issue still relevant for you?

**Comment by adam11grafik** [comment]
Hi,
I have similar requirement like @igorstojanovski.
I am working with test management tool like HP ALM and QTest.
I want to use my own engine from junit5 to change a little logic of discovering JUnit5 tests base on executed test set or test suite from test management.
I read carefully all comments here and tried with many options but did not achieved excluding junit-jupiter engine when we run build from maven.
Also it looks that maven-surefire does not support excluding engines. Or did I miss something?
I found excluding engines in Gradle but in our project we have to use Maven.
Can you advice something or add possibility to exclude engine/engines maybe via junit-platform.properties?
Best regards,
Adam
